### ISSUE TITLE
Spack on Cray KNL system and module detection error

### ISSUE THREAD

**Issue by pramodk** [LLNL/spack#2572]
Hello,

As posted on [google group](https://groups.google.com/forum/#!topic/spack/WdbtdkZb6lk), I am trying Spack on Cray KNL system. On other Cray systems (XK-7/XC-30), `spack compiler find` auto-detects **PrgEnv-***.  On this KNL system, this auto-detection isn't working.

I am looking at `lib/spack/spack/operating_systems/cnl.py`, line 56:

```
            output = modulecmd(
                'avail', cmp_cls.PrgEnv_compiler, output=str, error=str)

            # added this debug line
            print "DEBUG ===> : ", cmp_cls, paths, cmp_cls.PrgEnv_compiler, "OUTPUT : =>", output
```
With this, `spack compiler find` gives : 

```
DEBUG ===> :  <class 'spack.compilers.Gcc'> ('/home/kumbhar/spack/bin', '/opt/cray/pe/mpt/7.5.0......................................................................./misc', '/usr/lib/mit/bin', '/usr/lib/mit/sbin', '/opt/cray/pe/bin') gcc OUTPUT : => ModuleCmd_Avail.c(821):ERROR:64: Directory '/usr/bin/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/jdb' not found
utility.c(2263):ERROR:50: Cannot open file '/usr/bin/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/keygen' for 'reading'
ModuleCmd_Avail.c(821):ERROR:64: Directory '/usr/bin/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/X11/ControlPanel' not found
utility.c(2263):ERROR:50: Cannot open file
..............................................

```

Before looking into further details, anything obviously wrong here?

Thanks!

**Comment by tgamblin** [comment]
@mamelara @becker33

**Comment by mamelara** [comment]
Is the issue that it's not detecting all the compilers from modules? Just looking for clarification but from what I tried it looks like in cnl.py your $PATH is replacing $MODULEPATH.
```
            if paths:
                module_paths = ':' + ':'.join(p for p in paths)
                os.environ['MODULEPATH'] = module_paths
```
and it grabs $PATH from cmd/compiler.py in compiler_find:

```
    paths = args.add_paths
    if not paths:
        paths = get_path('PATH')
```

FYI I'm on Cori.

**Comment by becker33** [comment]
If the path is the issue, then this is fixed by #1514. That PR is only waiting on a rebase to current develop.

**Comment by KineticTheory** [comment]
@mamelara On Trinitite/Trinity, when I develop outside of spack, I target KNL instead of haswell by `module load PrgEnv-intel; module swap craype-haswell craype-mic-knl`.  The `spack compilers` command has no way of knowing about this `swap`.  How should I inform spack that I want to target haswell.

**Comment by becker33** [comment]
@KineticTheory that should be handled through the architecture interface, specifically the target. Installing with `spack install foo %intel target=mic-knl` should automatically load the craype-mic-knl module in the build environment.

**Comment by KineticTheory** [comment]
@becker33 That syntax doesn't seem to work for me.  Spack doesn't know the difference between the haswell's module set and the mic-knl's. Spack sees that I have already built for haswell, and refuses to build a mic-knl flavor.

```
spack install gsl@2.1 target=mic-knl
==> gsl is already installed in /usr/projects/draco/vendors/spack.tt.20161114/opt/spack/cray-CNL-haswell/gcc-5.3.0/gsl-2.1-d5okymwf627mlbgfgwfzafskdxbomjpp
```

**Comment by pramodk** [comment]
thanks @becker33 for pointing out #1514! The compilers (i.e. all **PrgEnv-***) are now getting populated. But there is an issue with the target in `compilers.yaml`:

```
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules:
    - PrgEnv-gnu
    - gcc/5.3.0
    operating_system: CNL
    paths:
      cc: cc
      cxx: CC
      f77: ftn
      fc: ftn
    spec: gcc@5.3.0
    target: <built-in function any>
```

And `spack arch` gives:

```
$ spack arch
cray-CNL-mic_knl
```

**Comment by becker33** [comment]
Can you check whether this is still reproducible with #2580 merged?

**Comment by becker33** [comment]
@pramodk Did you find #1514 useful? I notice that @robertdfrench closed his PR without it being merged.

**Comment by pramodk** [comment]
the issue of `target: <built-in function any>` fixed by #2580 merge.

We still need #1514 to automatically populate **PrgEnv-***.

**Comment by becker33** [comment]
@pramodk Would you be willing to reopen #1514 and rebase on current develop?

**Comment by mamelara** [comment]
@becker33 I rebased on current develop. Not sure how to reopen it as #1514

**Comment by becker33** [comment]
Push to the branch used for the PR (most PRs allow you limited access to their repo to affect only their PR branch). Then once the PR is up-to-date, reopen it.

**Comment by pramodk** [comment]
We can close this issue  once we merge #2745 (modules are now getting populated correctly). thanks @mamelara!

**Comment by becker33** [comment]
Closed by #2745
